File: TaskTracker/ClientApp/app/models/task.ts
import * as moment from "moment";
import { Moment, Duration } from "moment";

export type TaskStatus = 'active' | 'completed';

export type SortOrder = 'name' | 'priority' | 'added' | 'duration';

export class Task {
    public id: number;
    public name: string;
    public description: string;
    public priority: number;
    public added: Moment;
    public edited: Moment;
    public duration: Duration;
    public status?: TaskStatus;

    public static fromJson(json: any): Task {
        return <Task>{
            ...json,
            duration: moment.duration(json.duration),
            added: moment(json.added),
            edited: moment(json.edited),
        }
    }

    public static json(task: Task): any {
        return {
            ...task,
            duration: moment().startOf('day').add(task.duration).format("hh:mm:ss"),
            added: task.added.format("YYYY-MM-DD hh:mm:ss"),
            edited: task.edited.format("YYYY-MM-DD hh:mm:ss")
        }
    }
}

File: TaskTracker.Domain/Task.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System;
using System.ComponentModel.DataAnnotations;

namespace TaskTracker.Domain
{
    public class Task
    {
        public static Task Create(string name, string description, int priority, TimeSpan timeToComplete, DateTime added)
        {
            return new Task
            {
                Name = name,
                Description = description,
                Priority = priority,
                Added = added,
                Edited = added,
                Duration = timeToComplete,
                Status = TaskStatus.Active
            };
        }

        public Task()
        {
        }

        public int Id { get; set; }

        public string Name { get; set; }

        public string Description { get; set; }
        
        public int Priority { get; set; }

        [JsonConverter(typeof(DateFormatConverter), "yyyy-MM-dd hh:mm:ss")]
        public DateTime Added { get; set; }

        [ConcurrencyCheck]
        [JsonConverter(typeof(DateFormatConverter), "yyyy-MM-dd hh:mm:ss")]
        public DateTime Edited { get; set; }
        
        public TimeSpan Duration { get; set; }

        [JsonConverter(typeof(StringEnumConverter), true)]
        public TaskStatus Status { get; set; }
    }
}
